### PR TITLE
Automated builder schema tests for Kotlin builders and refactor

### PR DESCRIPTION
**Pull request by LukasPaczos**
Closes https://github.com/mapbox/mapbox-navigation-android/issues/2709 and closes https://github.com/mapbox/mapbox-navigation-android/issues/3369.

This PR builds on top of https://github.com/mapbox/mapbox-navigation-android/pull/3364 and adds an abstract `BuildeTest` class, that when extended by a regular test, runs a suite that aims to ensure that the builders are following the spec discussed in https://github.com/mapbox/mapbox-navigation-android/issues/2709#issuecomment-658869378.

We can consider this test suite as a mid-sized strainer - it will catch most of the issues that could be introduced, but there are still edge cases that are extremely hard to test in an automated fashion and we'll need to keep an eye on them when reviewing. One example could be `val` and  `var` definitions in Kotlin constructor - when access via reflection, they are listed as regular fields and we'd need to do a whole lot of checking and iterations to figure out that it could be a modifiable field passed through a constructor (and then having a separate setter method) which would be incorrect. This is mostly a concern for the new classes. Existing ones are still guarded by metalava which will give us insight when we're breaking something that wasn't caught by a test.

There's also a gotcha with JUint4 implementation - if a class doesn't have any JUnit `@Test` annotations, it will be ignored, even if the abstract parent does have tests. I'm working around that with the `trigger` method which **MUST HAVE** a `@Test` annotation in the overridden method as well. There are possible improvements here that could be made by creating a custom linter or a custom test runner, or maybe something else like a delegation pattern instead of inheritance.

This PR also re-enables `ktlint` for the UI module which has been disabled by accident in the repo refactor.

/cc @mapbox/navigation-android

### CODE REVIEW

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+            requiredValues.add(kProperty.getter.call(builderInstance)!!)
+        }
+
+        val optionalFieldValues = mutableListOf<Pair<KProperty<*>, Any>>()
+        (builderClass.members.filter { it is KProperty && it is KMutableProperty } as List<KProperty<*>>).forEachIndexed { index, kProperty ->
+            kProperty.isAccessible = true
+            optionalFieldValues.add(Pair(kProperty, kProperty.getter.call(builderInstance)!!))
+        }
+
+        optionalFieldValues.forEach { exclude ->
```
Comment (korshaknn) [inline libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:288]: val field = exclude.first
val value = exclude.second

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+ *          */
+ *         fun build(): Example {
+ *             return Example(
+ *                 required,
+ *                 foo,
+ *                 bar
+ *             )
+ *         }
+ *
+ *         override fun equals(other: Any?): Boolean {
```
[the remark this answers is not in the document]
Comment (LukasPaczos) [reply libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:104]: I didn't see a use-case tbh.

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+
+        val requiredFields =
+            builderClass.members.filter { it is KProperty && it !is KMutableProperty } as List<KProperty<*>>
+        val requiredValues = mutableListOf<Any>()
+        requiredFields.forEachIndexed { index, kProperty ->
+            kProperty.isAccessible = true
+            requiredValues.add(kProperty.getter.call(builderInstance)!!)
+        }
+
+        val optionalFieldValues = mutableListOf<Pair<KProperty<*>, Any>>()
```
[the remark this answers is not in the document]
Comment (LukasPaczos) [reply libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:282]: Required values don't have corresponding builder methods, they are always passed in the constructor. So I just take the values and unwrap them, I don't need to call any methods using the `KProperty` name.

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (korshaknn) [reply libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:104]: yep, so remove it from examples ^^

[the remark this answers is not in the document]
Comment (LukasPaczos) [reply libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:104]: oh, right :)

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+ *                 required,
+ *                 foo,
+ *                 bar
+ *             )
+ *         }
+ *     }
+ * }
+ * ```
+ */
+abstract class BuilderTest<Implementation : Any, Builder> {
```
Comment (kmadsen) [inline libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:107]: Did you try making a utility class where you pass it the parameters? When we use inheritance we can only inherit one piece of functionality

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+    private val implClass: KClass<*> by lazy { getImplementationClass() }
+
+    /**
+     * Your builder's resulting type.
+     */
+    abstract fun getImplementationClass(): KClass<Implementation>
+
+    /**
+     * Make sure that all of the required and optional values are provided and that they distinct from default ones.
```
Comment (kmadsen) [inline libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:117]:
```suggestion
     * Override this function with an expected builder with values.
```

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+     */
+    abstract fun getFilledUpBuilder(): Builder
```
Comment (kmadsen) [inline libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:119]: if we go with using inheritance, consider returning a list<Builder> here so we can test other scenarios. not sure about failure scenarios

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        builderClass.members.filter { it is KProperty && it !is KMutableProperty }.map { it.name }
+    private val optionalFieldNames =
+        builderClass.members.filter { it is KProperty && it is KMutableProperty }.map { it.name }
+    private val buildMethod = builderClass.members.find { it.name == "build" } as? KFunction
+        ?: throw RuntimeException("missing Builder.build method")
+
+    /**
+     * This method needs to be overridden and HAS TO explicitly add the @Test annotation in the child.
+     * This is needed to trick JUnit4 to process the test class if all the actual test cases are in the abstract parent.
+     * TODO add a lint rule to detect the missing annotation in the child, or create a new test runner
```
Comment (kmadsen) [inline libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:136]: What if this class was a TestRule? we could potentially eliminate the need for this

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+
+    @Test
+    fun impl_allFieldsAreVals() {
+        assertTrue(
+            implClass.members.filterIsInstance<KProperty<*>>().all { it !is KMutableProperty }
+        )
+    }
+
+    @Test
+    fun builder_allOptionalFieldsHaveCorrespondingImplMethod() {
```
Comment (kmadsen) [inline libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:170]: 🥇 🏅 💯

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+    }
+
+    @Test
+    fun hashCodeTest() {
+        compareOriginalNotEqualToRebuilt("hashCode") { instance ->
+            return@compareOriginalNotEqualToRebuilt instance.hashCode()
+        }
+    }
+
+    private fun compareOriginalNotEqualToRebuilt(
```
Comment (kmadsen) [inline libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:243]: 🥇 🏅 💯

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+
+        optionalFieldValues.forEach { exclude ->
+            val field = exclude.first
+            val value = exclude.second
+            val newBuilderInstance =
+                builderClass.constructors.first().call(*requiredValues.toTypedArray())
+            field.isAccessible = true
+            val defaultValue = field.getter.call(newBuilderInstance)
+            if (defaultValue == value) {
+                throw RuntimeException("make sure the provided value is different than default for \"${field.name}\"")
```
Comment (kmadsen) [inline libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:274]:
```suggestion
                throw RuntimeException("Make sure the getFilledUpBuilder() creates unique values for  \"${field.name}\": $defaultValue should not equal $value ")
```

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+            }
+            optionalFieldValues.filter { it != exclude }.forEach { fieldValue ->
+                (builderClass.members.find { it is KFunction && it.name == fieldValue.first.name }
+                    ?: throw RuntimeException("field name is not equal to method name for ${fieldValue.first.name} field"))
```
Comment (kmadsen) [inline libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:278]:
```suggestion
                    ?: throw RuntimeException("Make sure the ${fieldValue.first.name} is equal to the function name ${it.name}"))
```

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+ * ``` kotlin
+ * /**
+ *  * Description of the RequiredExample. This is a data class because it is responsible for
+ *  * transporting data values.
+ *  *
+ *  * @param required description of what to expect from required
+ *  * @param foo description of what to expect from foo
+ *  * @param bar description of what to expect from bar
+ *  */
+ * class Example private constructor(
```
Comment (kmadsen) [inline libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:30]: ❤️  really good idea to add the example here

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt (hunk 4 above)
Comment (LukasPaczos) [reply libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:107]: I did, but in the end, I wanted to make creating the test as easy as possible. If it was a utility class, we would still need to type out all of the test methods manually that then call utilities. Now we only need to type one to kickstart the test, the rest is run automatically.

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt (hunk 5 above)
Comment (LukasPaczos) [reply libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:117]: I'd prefer to leave as is.

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt (hunk 6 above)
Comment (LukasPaczos) [reply libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:119]: Currently, we are already iterating through all of the optional fields and testing equality for each one of them, so it's pretty exhaustive. I'm not sure how this could be automated when not all of the params are provided, we need all param instances anyway to perform the tests. I tried creating mocks using reflection, but because they are generic, I run into a lot of issues and didn't want to invest more time into it.

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt (hunk 7 above)
Comment (LukasPaczos) [reply libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:136]: I believe test rules can only perform actions before and after the test, they cannot run actual test cases. Also, they are applied before the test setup, so it would be too early to execute any assertions if there's a need for preparation.

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt (hunk 4 above)
Comment (kmadsen) [reply libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:107]: yeah that's fair, still happy with the test. also can see we are still testing failure scenarios https://github.com/mapbox/mapbox-navigation-android/pull/3382/files#diff-606a7ab2c9307301f681d1a1cfa57e94R57

File: libnavigation-base/src/test/java/com/mapbox/navigation/base/options/DeviceProfileTest.kt [hunk 13]
```diff
@@ -0,0 +1,21 @@
...
+import kotlin.reflect.KClass
+import org.junit.Test
+
+class DeviceProfileTest : BuilderTest<DeviceProfile, DeviceProfile.Builder>() {
+    override fun getImplementationClass(): KClass<DeviceProfile> = DeviceProfile::class
+
+    override fun getFilledUpBuilder(): DeviceProfile.Builder {
+        return DeviceProfile.Builder()
+            .customConfig("123")
+            .deviceType(mockk(relaxed = true))
```
Comment (kmadsen) [inline libnavigation-base/src/test/java/com/mapbox/navigation/base/options/DeviceProfileTest.kt:14]: nit: think we should send the non-default DeviceType here, it's a value type and doesn't have a builder covered by tests. enums probably classify as strings or ints

Comment (LukasPaczos) [reply libnavigation-base/src/test/java/com/mapbox/navigation/base/options/DeviceProfileTest.kt:14]: Ah it's an enum. We should be fine to compare equality though.

File: libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt (hunk 4 above)
Comment (RingerJK) [inline libtesting-utils/src/main/java/com/mapbox/navigation/testing/BuilderTest.kt:107]: ❤️

**Comment by RingerJK** [comment]
nice work!
